network/netconf/netconf_config.py

```diff
@@ -0,0 +1,229 @@
+
+'''
+
+import xml.dom.minidom
+try:
+    import ncclient.manager
+    HAS_NCCLIENT = True
+except ImportError:
+    HAS_NCCLIENT = False
+
+
+import logging
+
+
+def netconf_edit_config(m, xml, commit, retkwargs):
+    if ":candidate" in m.server_capabilities:
+        datastore = 'candidate'
+    else:
+        datastore = 'running'
+    m.lock(target=datastore)
+    try:
+        m.discard_changes()
+        config_before = m.get_config(source=datastore)
+        m.edit_config(target=datastore, config=xml)
+        config_after = m.get_config(source=datastore)
+        changed = config_before.data_xml != config_after.data_xml
+        if changed and commit:
+            if ":confirmed-commit" in m.server_capabilities:
+                m.commit(confirmed=True)
+                m.commit()
+            else:
+                m.commit()
+        return changed
+    finally:
+        m.unlock(target=datastore)
+
+
+# ------------------------------------------------------------------- #
+# Main
+
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            host=dict(type='str', required=True),
+            port=dict(type='int', default=830),
+            hostkey_verify=dict(type='bool', default=True),
+            username=dict(type='str', required=True, no_log=True),
+            password=dict(type='str', required=True, no_log=True),
+            operation=dict(choices=['edit_config'], default='edit_config'),
+            xml=dict(type='str', required=True),
+        )
+    )
+
+    if not HAS_NCCLIENT:
+        module.fail_json(msg='could not import the python library '
+                         'ncclient required by this module')
+
+    try:
+        xml.dom.minidom.parseString(module.params['xml'])
+    except:
+        e = get_exception()
+        module.fail_json(
+            msg='error parsing XML: ' +
+                str(e)
+        )
+        return
+
+    operation = module.params['operation']
+    nckwargs = dict(
+        host=module.params['host'],
+        port=module.params['port'],
+        hostkey_verify=module.params['hostkey_verify'],
+        username=module.params['username'],
+        password=module.params['password'],
+    )
+    retkwargs = dict()
+
+    try:
+        m = ncclient.manager.connect(**nckwargs)
+    except ncclient.transport.errors.AuthenticationError:
+        module.fail_json(
+            msg='authentication failed while connecting to device'
+        )
+    except:
+        e = get_exception()
+        module.fail_json(
+            msg='error connecting to the device: ' +
+                str(e)
+        )
+        return
+    retkwargs['server_capabilities'] = list(m.server_capabilities)
+    try:
+        changed = netconf_edit_config(
+            m=m,
+            xml=module.params['xml'],
+            commit=True,
+            retkwargs=retkwargs,
+        )
+    finally:
```

Comment:
I think using "try", "except" and "finally" together is incompatible (I'm guessing from what is written at https://docs.python.org/2.4/tut/node10.html and https://github.com/schacon/hg-git/issues/56). That is not the case here, as there is no "except"... Am I missing something?